File: 5-Problem-A-Day/154. Find Minimum in Rotated Sorted Array II.java
class Solution {
    public int findMin(int[] nums) {
        int begin = 0, end = nums.length - 1;
        while (begin < end) {
            int mid = begin + (end - begin) / 2;
            if (nums[mid] < nums[end])
                end = mid;
            else if (nums[mid] > nums[end])
                begin = mid + 1;
            else
                end -= 1;
        }
        return nums[begin];
    }
}

File: 2021/154. Find Minimum in Rotated Sorted Array II.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        int left = 0, right = nums.size() - 1;
        while (left < right) {            
            while (left < right && nums[left] == nums[left + 1]) {
                left++;
            }
            while (left < right && nums[right] == nums[right - 1]) {
                right--;
            }            
            int mid = left + (right - left) / 2;
            if (nums[mid] >= nums.back()) {
                left = mid + 1;
            } else {
                right = mid;
            }            
        }
        return left >= nums.size() ? nums.back() : nums[left];
    }
};

class Solution {
public:
    int findMin(vector<int>& nums) {
        int left = 0, right = (int) nums.size() - 1;
        while (left < right) {
            while (left < right && nums[left] == nums[left + 1]) {
                left++;
            }
            while (left < right && nums[right] == nums[right - 1]) {
                right--;
            }
            int mid = left + (right - left) / 2;
            if (nums[mid] > nums[right]) {
                left = mid + 1;
            } else {
                right = mid;
            }            
        }
        return nums[left];
    }
};